### PR TITLE
Feat/connection history

### PR DESCRIPTION
**Pull request by DorisWongNHS**
TISNEW-5793

### CODE REVIEW

File: src/main/java/uk/nhs/hee/tis/revalidation/connection/dto/ConnectionDto.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+@Builder
+@JsonIgnoreProperties(ignoreUnknown = true)
+@ApiModel(description = "List of connection information")
+public class ConnectionDto {
+  private List<ConnectionHistoryDto> connectionHistoryDtos;
+}
```
Comment (JosephKelly) [inline src/main/java/uk/nhs/hee/tis/revalidation/connection/dto/ConnectionDto.java:19]: This is interesting.  Is there a particular use case when referencing a `List<ConnectionHistoryDto>` doesn't work as well?

**Review by babul3012 — COMMENTED** [review]
How about programme history?

File: src/test/java/uk/nhs/hee/tis/revalidation/connection/controller/ConnectionControllerTest.java [hunk 2]
```diff
@@ -81,6 +107,41 @@ public void shouldRemoveDoctor() throws Exception {
         .andExpect(status().isOk());
   }
 
+  @Test
+  public void shouldReturnAllConnectionsForADoctor() throws Exception {
+    final var connectionDto = prepareConnectionDto();
+    when(connectionService.getTraineeConnectionInfo(gmcId)).thenReturn(connectionDto);
+    this.mockMvc.perform(get("/api/connections/{gmcId}", gmcId))
+        .andExpect(status().isOk())
+        .andDo(print())
```
Comment (JosephKelly) [inline src/test/java/uk/nhs/hee/tis/revalidation/connection/controller/ConnectionControllerTest.java:116]: If this is being used for manual verification of the result, please try to replace it with automated assertions.

File: src/test/java/uk/nhs/hee/tis/revalidation/connection/controller/ConnectionControllerTest.java [hunk 3]
```diff
@@ -81,6 +107,41 @@ public void shouldRemoveDoctor() throws Exception {
...
+        .andExpect(content().json(objectMapper.writeValueAsString(connectionDto)));
+  }
+
+  @Test
+  public void shouldNotFailWhenThereIsNoConnectionsForADoctor() throws Exception {
+    when(connectionService.getTraineeConnectionInfo(gmcId)).thenReturn(new ConnectionDto());
+    this.mockMvc.perform(get("/api/connections/{gmcId}", gmcId))
+        .andExpect(status().isOk())
+        .andDo(print());
```
Comment (JosephKelly) [inline src/test/java/uk/nhs/hee/tis/revalidation/connection/controller/ConnectionControllerTest.java:125]: If this is being used for manual verification of the result, please try to replace it with automated assertions.

**Review by JosephKelly — CHANGES_REQUESTED** [review]
There's an example of validating json results in https://github.com/Health-Education-England/tis-common-upload/blob/master/src/test/java/uk/nhs/hee/tis/common/upload/controller/AwsStorageControllerTest.java

**Review by uzairbaig — DISMISSED** [review]

File: src/test/java/uk/nhs/hee/tis/revalidation/connection/controller/ConnectionControllerTest.java (hunk 2 above)
Comment (DorisWongNHS) [reply src/test/java/uk/nhs/hee/tis/revalidation/connection/controller/ConnectionControllerTest.java]: It is updated, please have a look.

File: src/test/java/uk/nhs/hee/tis/revalidation/connection/controller/ConnectionControllerTest.java (hunk 3 above)
Comment (DorisWongNHS) [reply src/test/java/uk/nhs/hee/tis/revalidation/connection/controller/ConnectionControllerTest.java]: It is updated, please have a look. Thanks

**Comment by DorisWongNHS** [comment]
> How about programme history?

@babul3012 This will be done in the Integration Service to join the endpoints.

File: src/main/java/uk/nhs/hee/tis/revalidation/connection/dto/ConnectionDto.java (hunk 1 above)
Comment (DorisWongNHS) [reply src/main/java/uk/nhs/hee/tis/revalidation/connection/dto/ConnectionDto.java:19]: It was also working. But was discussed with the team, and changed it to keep the practice for the frontend to consume and keep it flexible if we need to add count/total results in the future.

**Review by jayanta2018 — DISMISSED** [review]

**Review by babul3012 — APPROVED** [review]

**Review by jayanta2018 — DISMISSED** [review]

**Review by babul3012 — DISMISSED** [review]

**Review by JosephKelly — APPROVED** [review]
Good stuff! Greatly improved the code.

**Pull request merged**